File: lesson-04/java/ex02.java
import java.math.BigInteger;

public class ex02 {
    public static long powRec(long a, long b) {
        if (b == 0)
            return 1;
        else if (b % 2 == 1)
            return a * powRec(a, b - 1);
        else
            return powRec(a * a, b / 2);
    }

    public static BigInteger powWhile(BigInteger a, BigInteger b) {
        BigInteger result = new BigInteger("1");

        while (b.compareTo(BigInteger.ZERO) > 0) {
            if (b.mod(BigInteger.TWO).equals(BigInteger.ONE)) {
                result = result.multiply(a);
            }
            a = a.multiply(a);
            b = b.divide(BigInteger.TWO);
        }
        return result;
    }

    public static void main(String[] args) {
        System.out.println(
                powWhile(new java.math.BigInteger("2"),
                        new java.math.BigInteger("10000")));
    }
}

File: lesson-04/js/ex02.js
const powRec = (a, b) => {
    if (b === 0) return 1;
    else if (b % 2 === 1) return a * powRec(a, b - 1);
    else return powRec(a * a, Math.floor(b / 2));
}

const powWhile = (a, b) => {
    let result = BigInt(1);

    while (b > 0) {
        if (Number(b) % 2 === 1) {
            result = result * a;
        }
        a *= a;
        b = Math.floor(Number(b) / 2);
    }
    return result;
}
// BigInt
let a = BigInt(1);
console.log(typeof a);
// console.log(powRec(2, 100000));
console.log(powWhile(BigInt(2), BigInt(10000)));